File: BackEnd/src/main/java/com/example/demo/model/Task.java
package com.example.demo.model;

import java.time.Instant;
import java.util.List;
import java.util.UUID;

import org.springframework.data.annotation.CreatedDate;
import org.springframework.data.annotation.Id;
import org.springframework.data.annotation.LastModifiedDate;
import org.springframework.data.mongodb.core.index.Indexed;
import org.springframework.data.mongodb.core.mapping.Document;
import org.springframework.data.mongodb.core.mapping.Field;

@Document(collection = "tasks")
public class Task {
    @Id
    private UUID id;
    private String title;
    private String description;
    @Indexed
    @Field("projectId")
    private UUID projectId; // Tham chiếu Project
    @Indexed
    @Field("columnId")
    private UUID columnId; // Tham chiếu Column
    @Field("creatorId")
    private UUID creatorId; // Tham chiếu User
    @Indexed // Tối ưu hóa cho "My Tasks"
    private List<UUID> assignees; // Mảng các ID của User
    @Indexed
    @Field("dueDate")
    private Instant dueDate;
    private List<UUID> labels; // Mảng các ID của Label
    private List<Checklist> checklists;
    @CreatedDate
    @Field("createdAt")
    private Instant createdAt;
    @LastModifiedDate
    @Field("updatedAt")
    private Instant updatedAt;
}

File: backend/src/models/Task.js
// src/models/Task.js
const mongoose = require('mongoose');

const ChecklistItemSchema = new mongoose.Schema({
    _id: {
        type: mongoose.Schema.Types.UUID,
        default: () => new mongoose.Types.UUID(),
    },
    text: {
        type: String,
        required: true,
    },
    isCompleted: {
        type: Boolean,
        default: false,
    },
}, {
    _id: false, // _id của subdoc đôi khi gây rối, nhưng nếu cần quản lý item thì để true.
    // Nếu để _id thì cần toJSON ở đây nữa, nhưng đơn giản hóa thì ta xử lý ở parent
});

const TaskSchema = new mongoose.Schema({
    _id: {
        type: mongoose.Schema.Types.UUID,
        default: () => new mongoose.Types.UUID(),
    },
    title: {
        type: String,
        required: [true, 'Please add a task title'],
        trim: true,
    },
    description: {
        type: String,
    },
    projectId: {
        type: mongoose.Schema.Types.UUID,
        ref: 'Project',
        required: true,
        index: true,
    },
    columnId: {
        type: mongoose.Schema.Types.UUID,
        ref: 'Column',
        required: true,
        index: true,
    },
    creatorId: {
        type: mongoose.Schema.Types.UUID,
        ref: 'User',
        required: true,
    },
    assignees: [{
        type: mongoose.Schema.Types.UUID,
        ref: 'User',
    }],
    dueDate: {
        type: Date,
        index: true,
    },
    labels: [{
        type: mongoose.Schema.Types.UUID,
        ref: 'Label',
    }],
    checklists: [ChecklistItemSchema],
}, {
    timestamps: true,
    collection: 'tasks',
    // --- THÊM PHẦN NÀY ---
    toJSON: {
        getters: true,
        virtuals: true,
        transform: (doc, ret) => {
            delete ret.__v;
            delete ret.id;

            // Convert các trường đơn lẻ
            const uuidFields = ['_id', 'projectId', 'columnId', 'creatorId'];
            uuidFields.forEach(field => {
                if (ret[field] && typeof ret[field] === 'object' && ret[field].toString) {
                    ret[field] = ret[field].toString();
                }
            });

            // Convert các trường mảng (Assignees, Labels)
            const arrayUUIDFields = ['assignees', 'labels'];
            arrayUUIDFields.forEach(field => {
                if (ret[field] && Array.isArray(ret[field])) {
                    ret[field] = ret[field].map(item => {
                        // Nếu là object đã populate (User/Label object) thì giữ nguyên hoặc xử lý riêng
                        // Nếu là Buffer/UUID raw thì convert
                        if (item && typeof item === 'object' && item.constructor.name === 'Binary') {
                            return item.toString();
                        }
                        return item;
                    });
                }
            });

            return ret;
        }
    }
});

module.exports = mongoose.model('Task', TaskSchema);